### ISSUE TITLE
feat(schematics): specify multiple default collections

### ISSUE THREAD

**Issue by jeffbcross** [angular/devkit#249]
### Bug Report or Feature Request (mark with an `x`)
```
- [ ] bug report -> please search issues before submitting
- [x] feature request
```

### Area
```
- [] devkit
- [x] schematics
```

### Desired functionality
It'd be nice to have multiple collections configured in my CLI project instead of a single default collection. I.e. I'd like to have a project that gives preference to `nrwl/schematics` but will fall back to `angular/schematics` if I try to generate something not provided by `nrwl/schematics`. This would make it easier to publish many focused schematics without requiring the user to specify the collection each time they're generating something new. My naive suggestion is to change the `.angular-cli.json` schema from a single `collection` value to a cascading array of `collections` like so:

```json
"defaults": {
  "collection": "nrwl/schematics"
}
```

```json
"defaults": {
  "collections": [
    "nrwl/schematics",
    "angular/schematics"
  ]
}
```

Then when calling `ng generate`, schematics would use the first collection that provides a matching schematic.

This issue seems cross-cutting between CLI and Schematics, so I thought this repository was the most appropriate place to open it, but lmk if I should open it on angular/angular-cli.

(This may seem similar to #34, but it's more focused on end user experience than schematics author experience).

**Comment by tomastrajan** [comment]
Hey guys!

I was able to implement this in @angular/cli exclusively and was rather a small change. Basically just retrieving of schematics property as array and filtering for first matching schematics / collection pair.

Still have to look into tests though. Also it is possible that you might have had plan for other solution so please let me know.

Also check following screens for the new behavior.

Display help for `ng generate` with all available collections and their schematics

![schema1](https://user-images.githubusercontent.com/3764868/33513048-737f5742-d78f-11e7-9f3f-23cbe069464f.png)

Display detailed help for particular schematics (first matching schematics / collection pair wins so the order in schematics property in `angular-cli.json` matters)

![schema2](https://user-images.githubusercontent.com/3764868/33513049-73e5c234-d78f-11e7-9a2d-b621b9308b25.png)

Generate items using schematics from different collections (again, first matching schematics / collection pair wins so the order in schematics property in `angular-cli.json` matters)

![schema3](https://user-images.githubusercontent.com/3764868/33513050-74175704-d78f-11e7-8ac0-08f3897fdcac.png)

**Comment by tomastrajan** [comment]
Hi @jeffbcross @hansl 

There is now open PR with initial working implementation of this feature on `angular/angular-cli` repo. Please let me know if its a good base to make this work.

Cheers 🍺

**Comment by myspivey** [comment]
Would like this as well.

**Comment by Hotell** [comment]
any update on this? 

this is totally needed for great DX. 

Enforcing users to write `--collection @my-foo-bar/super-schematics` all the time when switching between collections is super annoying...

thx

**Comment by tomastrajan** [comment]
From what I have seen in `nrwl/schematics` currently the way to go is to extend `@schematics/angular` in your custom `collection.json`. 

That way you don't have to delegate every schematic manualy, provide default ones out of the box and add custom on top of that.

Thr big downside is that this leads to ONE custom collection set as a new default for `angular-cli.json`. 

This unfortunatelly doesn't support "multiple small focused collections from different authors" use case. 😞

**Comment by Hotell** [comment]
Yeah I know, but this is to restrictive and repetitive and hard to maintain overall.  

Maybe, just guys, can you please give some statement or so, why @tomastrajan PR  wasn't accepted yet or what are the plans ? I know we are all super busy, but would appreciate some response. 

thank you 🖖

**Comment by clydin** [comment]
schematic extension supports 1 or more schematics with the last taking precedence.  The currently envisioned concept is for a custom collection to be placed within the project and that collection to extend from any other schematic collections with the option to directly define other custom schematics if desired.  As the schematic extension logic is fully implemented within the core of the schematics system, having similar but repeated logic within the CLI itself represents a maintenance and supportability concern.

**Comment by clydin** [comment]
For reference, the collection extension PR: https://github.com/angular/devkit/pull/398

**Comment by hansl** [comment]
> custom collection in the project

You can already do that.

**Comment by james-schwartzkopf** [comment]
Yeah, but there's none there by default in a project created by ng new.  So everyone's adding `@schematics/angular` to their schematic collections since it doesn't seen to be straightforward how you add a collection without having to use --collection either for the new collection or for `@schematics/angular`.

**Comment by alexeagle** [comment]
This issue was moved to angular/angular-cli#12157
